File: day_01/preprocess.py
import argparse
import sys

# It looks like that 'oneight' is equal to '1.*8' not '1ight'.


def replace_number_letters(s):
    n_letters = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

    cands = list(
        filter(
            lambda x: x[0] != -1,
            map(lambda x: [s.find(x[1]), x[0], x[1]], enumerate(n_letters, 1)),
        )
    )

    if len(cands) == 0:
        return s
    else:
        target_num, target_str = min(cands, key=lambda x: x[0])[1:]
        return replace_number_letters(
            s.replace(
                target_str,
                str(target_num) + target_str[1:],  # Please see the above comment.
                1,
            )
        )


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('infile', nargs='?', type=argparse.FileType(), default=sys.stdin)
    args = parser.parse_args()
    with args.infile as f:
        for line in f:
            print(replace_number_letters(line), end='')

File: day_01/preprocess.pl
#!/usr/bin/env perl

# It looks like that 'oneight' is equal to '1.*8' not '1ight'.

use strict;
use warnings;

sub replace_number_letters {
    my $s = shift(@_);
    my @n_letters = qw(one two three four five six seven eight nine);
    
    my @cands = sort {$a->[0] <=> $b->[0]} grep {$_->[0] != -1} map {[index($s, $n_letters[$_]), $n_letters[$_], $_ + 1]} 0..$#n_letters;

    if (!@cands) {
        return $s;
    } else {
        my $offset = $cands[0][0];
        my $num = $cands[0][2];
        substr($s, $offset, 1, "$num");
        return replace_number_letters($s);
    }
}

while (<>) {
    chomp;
    print replace_number_letters($_) . "\n";
}
